File: PreStage-Mac/Sources/PreStage/Services/CopyService.swift
import Foundation
import CryptoKit

enum CopyServiceError: Error {
    case missingTarget
}

actor CopyOperationControl {
    private var paused = false
    private var cancelled = false

    func pause() {
        paused = true
    }

    func resume() {
        paused = false
    }

    func cancel() {
        cancelled = true
        paused = false
    }

    func isPaused() -> Bool {
        paused
    }

    func isCancelled() -> Bool {
        cancelled
    }

    func waitIfPaused() async -> Bool {
        while paused && !cancelled {
            try? await Task.sleep(nanoseconds: 120_000_000)
        }
        return !cancelled
    }
}

final class CopyService {
    func copyItems(
        _ items: [MediaItem],
        to targetFolder: URL,
        sourceRoot: URL?,
        rule: CopyOrganizationRule,
        conflictPolicy: CopyConflictPolicy,
        verificationMode: CopyVerificationMode = .sizeOnly,
        control: CopyOperationControl,
        progress: @escaping @MainActor (CopyProgress, UUID?, CopyStatus?) -> Void
    ) async {
        let totalBytes = items.reduce(Int64(0)) { $0 + $1.fileSize }
        var snapshot = CopyProgress(currentItem: "", completedCount: 0, totalCount: items.count, completedBytes: 0, totalBytes: totalBytes, isRunning: true, message: "Copying")
        await progress(snapshot, nil, nil)
        for item in items {
            await progress(snapshot, item.id, .queued)
        }

        for (index, item) in items.enumerated() {
            if await control.isPaused() {
                snapshot.isPaused = true
                snapshot.message = "Copy paused"
                await progress(snapshot, nil, nil)
            }
            guard await control.waitIfPaused() else {
                snapshot.isRunning = false
                snapshot.isCancelled = true
                snapshot.message = "Copy cancelled: \(snapshot.completedCount) of \(snapshot.totalCount) complete"
                for cancelledItem in items[index...] {
                    await progress(snapshot, cancelledItem.id, .cancelled)
                }
                await progress(snapshot, nil, nil)
                return
            }
            snapshot.isPaused = false

            await progress(snapshot, item.id, .copying)
            let folder = destinationFolder(for: item, targetFolder: targetFolder, sourceRoot: sourceRoot, rule: rule)

            do {
                try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
                if let destination = try destinationURL(for: item, in: folder, policy: conflictPolicy) {
                    if conflictPolicy == .overwrite, FileManager.default.fileExists(atPath: destination.path) {
                        try FileManager.default.removeItem(at: destination)
                        try removeSidecarIfNeeded(for: destination)
                    }

                    try FileManager.default.copyItem(at: item.url, to: destination)
                    try copySidecarIfNeeded(for: item.url, to: destination, conflictPolicy: conflictPolicy)
                    await progress(snapshot, item.id, .copied)
                    let status = try verificationStatus(for: item, destination: destination, mode: verificationMode)
                    await progress(snapshot, item.id, status)
                } else {
                    await progress(snapshot, item.id, .skipped)
                }
            } catch {
                await progress(snapshot, item.id, .failed)
            }

            snapshot.completedCount += 1
            snapshot.completedBytes += item.fileSize
            snapshot.currentItem = item.filename
            snapshot.message = "\(snapshot.completedCount) of \(snapshot.totalCount) complete"
            await progress(snapshot, nil, nil)

            if await control.isCancelled() {
                snapshot.isRunning = false
                snapshot.isCancelled = true
                snapshot.message = "Copy cancelled: \(snapshot.completedCount) of \(snapshot.totalCount) complete"
                for cancelledItem in items.dropFirst(index + 1) {
                    await progress(snapshot, cancelledItem.id, .cancelled)
                }
                await progress(snapshot, nil, nil)
                return
            }
        }

        snapshot.isRunning = false
        snapshot.message = "Finished: \(snapshot.completedCount) files checked by \(verificationMode.completionDescription)"
        await progress(snapshot, nil, nil)
    }

    private func verificationStatus(for item: MediaItem, destination: URL, mode: CopyVerificationMode) throws -> CopyStatus {
        let copiedSize = (try? destination.resourceValues(forKeys: [.fileSizeKey]).fileSize).map(Int64.init) ?? -1
        guard copiedSize == item.fileSize else { return .failed }

        switch mode {
        case .sizeOnly:
            return .verified
        case .sha256:
            let sourceHash = try sha256Digest(for: item.url)
            let destinationHash = try sha256Digest(for: destination)
            return sourceHash == destinationHash ? .verified : .failed
        }
    }

    private func sha256Digest(for url: URL) throws -> SHA256.Digest {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }

        var hasher = SHA256()
        while true {
            let data = try handle.read(upToCount: 1_048_576) ?? Data()
            guard !data.isEmpty else { break }
            hasher.update(data: data)
        }
        return hasher.finalize()
    }

    private func destinationURL(for item: MediaItem, in folder: URL, policy: CopyConflictPolicy) throws -> URL? {
        let baseURL = folder.appendingPathComponent(item.filename)
        guard mediaOrSidecarExists(for: baseURL) else { return baseURL }

        switch policy {
        case .skipExisting:
            return nil
        case .overwrite:
            return baseURL
        case .autoRename:
            return nextAvailableURL(for: baseURL)
        }
    }

    private func nextAvailableURL(for url: URL) -> URL {
        let folder = url.deletingLastPathComponent()
        let baseName = url.deletingPathExtension().lastPathComponent
        let ext = url.pathExtension

        var index = 2
        while true {
            let filename = ext.isEmpty ? "\(baseName) \(index)" : "\(baseName) \(index).\(ext)"
            let candidate = folder.appendingPathComponent(filename)
            if !mediaOrSidecarExists(for: candidate) {
                return candidate
            }
            index += 1
        }
    }

    private func mediaOrSidecarExists(for mediaURL: URL) -> Bool {
        if FileManager.default.fileExists(atPath: mediaURL.path) {
            return true
        }
        let sidecar = mediaURL.deletingPathExtension().appendingPathExtension("xmp")
        return FileManager.default.fileExists(atPath: sidecar.path)
    }

    private func removeSidecarIfNeeded(for mediaURL: URL) throws {
        let sidecar = mediaURL.deletingPathExtension().appendingPathExtension("xmp")
        guard FileManager.default.fileExists(atPath: sidecar.path) else { return }
        try FileManager.default.removeItem(at: sidecar)
    }

    private func copySidecarIfNeeded(for sourceMediaURL: URL, to destinationMediaURL: URL, conflictPolicy: CopyConflictPolicy) throws {
        let sourceSidecar = sourceMediaURL.deletingPathExtension().appendingPathExtension("xmp")
        guard FileManager.default.fileExists(atPath: sourceSidecar.path) else { return }
        let destinationSidecar = destinationMediaURL.deletingPathExtension().appendingPathExtension("xmp")
        if FileManager.default.fileExists(atPath: destinationSidecar.path) {
            guard conflictPolicy == .overwrite else { return }
            try FileManager.default.removeItem(at: destinationSidecar)
        }

        if conflictPolicy == .autoRename, FileManager.default.fileExists(atPath: destinationSidecar.path) {
            return
        }
        try FileManager.default.copyItem(at: sourceSidecar, to: destinationSidecar)
    }

    private func destinationFolder(for item: MediaItem, targetFolder: URL, sourceRoot: URL?, rule: CopyOrganizationRule) -> URL {
        switch rule {
        case .captureDate:
            let day = AppFormatters.dayFolderFormatter.string(from: item.captureDate ?? item.createdDate ?? item.modifiedDate ?? Date())
            return targetFolder.appendingPathComponent(day, isDirectory: true)
        case .preserveStructure:
            guard let sourceRoot else { return targetFolder }
            let sourceFolder = item.url.deletingLastPathComponent()
            let rootPath = sourceRoot.standardizedFileURL.path
            let sourcePath = sourceFolder.standardizedFileURL.path
            guard sourcePath.hasPrefix(rootPath) else { return targetFolder }
            let relativePath = String(sourcePath.dropFirst(rootPath.count)).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            guard !relativePath.isEmpty else { return targetFolder }
            return targetFolder.appendingPathComponent(relativePath, isDirectory: true)
        case .cameraModel:
            return targetFolder.appendingPathComponent(safeFolderName(item.cameraModel ?? "Unknown Camera"), isDirectory: true)
        case .rating:
            return targetFolder.appendingPathComponent("\(item.rating)-Star", isDirectory: true)
        }
    }

    private func safeFolderName(_ value: String) -> String {
        let invalid = CharacterSet(charactersIn: "/:")
        let parts = value.components(separatedBy: invalid).filter { !$0.isEmpty }
        return parts.joined(separator: "-").trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

File: PreStage.Core/Services/CopyService.cs
using System.Security.Cryptography;
using PreStage.Core.Models;

namespace PreStage.Core.Services;

public class CopyService
{
    private readonly object _pauseLock = new();
    private bool _paused;
    private CancellationTokenSource? _cts;

    public CopyProgress Progress { get; } = new();

    public event Action<string>? OnLogMessage;

    public async Task<CopyLogRecord> CopyAsync(
        List<MediaItem> items,
        string targetFolder,
        CopyOrganizationRule rule,
        CopyConflictPolicy conflictPolicy,
        CopyContentMode contentMode,
        CopyVerificationMode verificationMode,
        CancellationToken ct = default,
        string? sourceRoot = null)
    {
        _cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        var token = _cts.Token;

        Progress.Reset();
        Progress.IsRunning = true;

        var filtered = contentMode == CopyContentMode.RawOnly
            ? items.Where(i => i.MediaType == MediaType.Raw).ToList()
            : items.ToList();

        Progress.TotalCount = filtered.Count;
        Progress.TotalBytes = filtered.Sum(i => i.FileSize);
        Progress.Message = "Starting copy...";

        var log = new CopyLogRecord
        {
            SourcePath = filtered.FirstOrDefault()?.Url,
            DestinationPath = targetFolder,
            Rule = rule,
            TotalItems = filtered.Count,
            TotalBytes = filtered.Sum(i => i.FileSize),
            Entries = []
        };

        var copiedCount = 0;
        var copiedBytes = 0L;

        foreach (var item in filtered)
        {
            token.ThrowIfCancellationRequested();
            await WaitIfPaused(token);

            Progress.CurrentItem = item.Filename;
            Progress.Message = $"Copying {item.Filename}...";

            var destPath = BuildDestinationPath(item, targetFolder, rule, sourceRoot);
            Directory.CreateDirectory(Path.GetDirectoryName(destPath)!);

            var entry = new CopyLogEntry
            {
                Filename = item.Filename,
                Timestamp = DateTime.UtcNow
            };

            try
            {
                if (MediaOrSidecarExists(destPath))
                {
                    switch (conflictPolicy)
                    {
                        case CopyConflictPolicy.SkipExisting:
                            entry.Status = CopyStatus.Skipped;
                            entry.Message = "File already exists, skipped";
                            Log($"Skipped: {item.Filename}");
                            break;
                        case CopyConflictPolicy.Overwrite:
                            RemoveDestinationSidecar(destPath);
                            File.Copy(item.Url, destPath, true);
                            CopySidecarIfNeeded(item.Url, destPath, overwrite: true);
                            entry.Status = CopyStatus.Copied;
                            entry.Message = "Overwritten";
                            break;
                        case CopyConflictPolicy.AutoRename:
                            destPath = GetUniquePath(destPath);
                            File.Copy(item.Url, destPath);
                            CopySidecarIfNeeded(item.Url, destPath, overwrite: false);
                            entry.Status = CopyStatus.Copied;
                            entry.Message = $"Renamed to {Path.GetFileName(destPath)}";
                            break;
                    }
                }
                else
                {
                    File.Copy(item.Url, destPath);
                    CopySidecarIfNeeded(item.Url, destPath, overwrite: false);
                    entry.Status = CopyStatus.Copied;
                    entry.Message = "Copied";
                }

                if (entry.Status == CopyStatus.Copied)
                {
                    await VerifyFileAsync(item, destPath, verificationMode, entry);
                }
            }
            catch (OperationCanceledException)
            {
                entry.Status = CopyStatus.Cancelled;
                entry.Message = "Copy cancelled";
                log.Entries.Add(entry);
                Progress.IsCancelled = true;
                break;
            }
            catch (Exception ex)
            {
                entry.Status = CopyStatus.Failed;
                entry.Message = ex.Message;
                Log($"Failed: {item.Filename} - {ex.Message}");
            }

            if (entry.Status == CopyStatus.Copied || entry.Status == CopyStatus.Verified)
            {
                copiedCount++;
                copiedBytes += item.FileSize;
            }

            log.Entries.Add(entry);
            Progress.CompletedCount = copiedCount;
            Progress.CompletedBytes = copiedBytes;

            await Task.Delay(1, token);
        }

        Progress.IsRunning = false;
        log.FinishedAt = DateTime.UtcNow;

        if (!Progress.IsCancelled)
            Progress.Message = $"Copy complete: {copiedCount} files, {FormatBytes(copiedBytes)}";
        else
            Progress.Message = $"Copy cancelled: {copiedCount} files copied before cancel";

        return log;
    }

    public void Pause()
    {
        lock (_pauseLock) { _paused = true; }
        Progress.IsPaused = true;
        Progress.Message = "Paused";
    }

    public void Resume()
    {
        lock (_pauseLock)
        {
            _paused = false;
            Monitor.PulseAll(_pauseLock);
        }
        Progress.IsPaused = false;
        Progress.Message = "Resuming...";
    }

    public void Cancel()
    {
        _cts?.Cancel();
        Progress.IsCancelled = true;
        Progress.Message = "Cancelling...";
    }

    private async Task WaitIfPaused(CancellationToken ct)
    {
        while (true)
        {
            bool isPaused;
            lock (_pauseLock) { isPaused = _paused; }

            if (!isPaused) return;

            await Task.Run(() =>
            {
                lock (_pauseLock)
                {
                    while (_paused) Monitor.Wait(_pauseLock);
                }
            }, ct);
        }
    }

    private static string BuildDestinationPath(
        MediaItem item,
        string targetFolder,
        CopyOrganizationRule rule,
        string? sourceRoot)
    {
        return rule switch
        {
            CopyOrganizationRule.PreserveStructure => BuildPreservePath(item, targetFolder, sourceRoot),
            CopyOrganizationRule.CameraModel => BuildCameraPath(item, targetFolder),
            CopyOrganizationRule.Rating => BuildRatingPath(item, targetFolder),
            _ => BuildDatePath(item, targetFolder)
        };
    }

    private static string BuildDatePath(MediaItem item, string targetFolder)
    {
        var date = item.CaptureDate ?? item.CreatedDate ?? DateTime.Now;
        var subFolder = date.ToString("yyyy-MM-dd");
        return Path.Combine(targetFolder, subFolder, item.Filename);
    }

    private static string BuildPreservePath(MediaItem item, string targetFolder, string? sourceRoot)
    {
        if (string.IsNullOrEmpty(item.Url) || string.IsNullOrEmpty(targetFolder))
            return Path.Combine(targetFolder, item.Filename);

        var sourceDir = Path.GetDirectoryName(item.Url) ?? "";
        if (!string.IsNullOrWhiteSpace(sourceRoot))
        {
            try
            {
                var rootFullPath = Path.GetFullPath(sourceRoot);
                var sourceDirFullPath = Path.GetFullPath(sourceDir);
                var relative = Path.GetRelativePath(rootFullPath, sourceDirFullPath);
                if (!relative.StartsWith("..", StringComparison.Ordinal) &&
                    !Path.IsPathRooted(relative))
                {
                    return relative == "."
                        ? Path.Combine(targetFolder, item.Filename)
                        : Path.Combine(targetFolder, relative, item.Filename);
                }
            }
            catch
            {
            }
        }

        var folderName = Path.GetFileName(sourceDir);
        return string.IsNullOrWhiteSpace(folderName)
            ? Path.Combine(targetFolder, item.Filename)
            : Path.Combine(targetFolder, folderName, item.Filename);
    }

    private static string BuildCameraPath(MediaItem item, string targetFolder)
    {
        var camera = string.IsNullOrWhiteSpace(item.CameraModel)
            ? "Unknown" : SanitizeFolderName(item.CameraModel);
        var date = item.CaptureDate ?? item.CreatedDate ?? DateTime.Now;
        var subFolder = $"{camera}/{date:yyyy-MM-dd}";
        return Path.Combine(targetFolder, subFolder, item.Filename);
    }

    private static string BuildRatingPath(MediaItem item, string targetFolder)
    {
        var rating = item.Rating > 0 ? $"{item.Rating}_stars" : "Unrated";
        return Path.Combine(targetFolder, rating, item.Filename);
    }

    private static string GetUniquePath(string path)
    {
        var dir = Path.GetDirectoryName(path) ?? "";
        var name = Path.GetFileNameWithoutExtension(path);
        var ext = Path.GetExtension(path);
        var counter = 1;

        string candidate;
        do
        {
            candidate = Path.Combine(dir, $"{name}_{counter}{ext}");
            counter++;
        } while (MediaOrSidecarExists(candidate));

        return candidate;
    }

    private static bool MediaOrSidecarExists(string mediaPath)
    {
        return File.Exists(mediaPath) || File.Exists(XmpService.GetSidecarPath(mediaPath));
    }

    private static void RemoveDestinationSidecar(string mediaPath)
    {
        var sidecarPath = XmpService.GetSidecarPath(mediaPath);
        if (File.Exists(sidecarPath))
            File.Delete(sidecarPath);
    }

    private static void CopySidecarIfNeeded(string sourceMediaPath, string destinationMediaPath, bool overwrite)
    {
        var sourceSidecar = XmpService.GetSidecarPath(sourceMediaPath);
        if (!File.Exists(sourceSidecar))
            return;

        var destinationSidecar = XmpService.GetSidecarPath(destinationMediaPath);
        File.Copy(sourceSidecar, destinationSidecar, overwrite);
    }

    private async Task VerifyFileAsync(MediaItem source, string destPath,
        CopyVerificationMode mode, CopyLogEntry entry)
    {
        var destInfo = new FileInfo(destPath);
        if (!destInfo.Exists)
        {
            entry.Status = CopyStatus.Failed;
            entry.Message = "Destination file not found after copy";
            return;
        }

        if (destInfo.Length != source.FileSize)
        {
            entry.Status = CopyStatus.Failed;
            entry.Message = $"Size mismatch: src={source.FileSize} dst={destInfo.Length}";
            return;
        }

        if (mode == CopyVerificationMode.Sha256)
        {
            var srcHash = await ComputeSha256Async(source.Url);
            var dstHash = await ComputeSha256Async(destPath);

            if (!string.Equals(srcHash, dstHash, StringComparison.OrdinalIgnoreCase))
            {
                entry.Status = CopyStatus.Failed;
                entry.Message = "SHA-256 mismatch";
                return;
            }
        }

        entry.Status = CopyStatus.Verified;
        entry.Message = mode == CopyVerificationMode.Sha256
            ? "Verified (size + SHA-256)" : "Verified (size)";
    }

    private static async Task<string> ComputeSha256Async(string path)
    {
        using var stream = File.OpenRead(path);
        var hash = await SHA256.HashDataAsync(stream);
        return Convert.ToHexStringLower(hash);
    }

    private void Log(string message)
    {
        OnLogMessage?.Invoke(message);
    }

    private static string FormatBytes(long bytes)
    {
        if (bytes < 1024) return $"{bytes} B";
        if (bytes < 1024 * 1024) return $"{bytes / 1024.0:F1} KB";
        if (bytes < 1024 * 1024 * 1024) return $"{bytes / (1024.0 * 1024):F1} MB";
        return $"{bytes / (1024.0 * 1024 * 1024):F2} GB";
    }

    private static string SanitizeFolderName(string name)
    {
        var invalid = Path.GetInvalidFileNameChars();
        foreach (var c in invalid) name = name.Replace(c, '_');
        return name.Trim();
    }
}
